processing/src/main/java/io/druid/query/lookup/LookupReferencesManager.java

```diff
@@ -98,287 +167,247 @@ public void start()
-          continue;
-        }
-        if (!lookupExtractorFactory.start()) {
-          failedExtractorFactoryMap.put(lookupName, lookupExtractorFactory);
-          continue;
-        }
-        if (null != lookupMap.putIfAbsent(lookupName, lookupExtractorFactory)) {
-          // handle race
-          LOGGER.warn("lookup [%s] is not add, another lookup with the same name already exist", lookupName);
-          if (!lookupExtractorFactory.close()) {
-            LOGGER.error("Failed to properly close stale lookup [%s]", lookupExtractorFactory);
-          }
-          continue;
-        }
-        if (lookupSnapshotTaker != null) {
-          lookupSnapshotTaker.takeSnapshot(getAllAsList());
-        }
+      if (!queue.offer(new DropNotice(lookupName), 1, TimeUnit.MILLISECONDS)) {
+        throw new ISE("notice queue add timedout to add [%s] lookup drop notice", lookupName);
       }
-      if (!failedExtractorFactoryMap.isEmpty()) {
-        throw new ISE(
-            "was not able to start the following lookup(s) [%s]",
-            failedExtractorFactoryMap.keySet().toString()
-        );
-      }
-    }
-    finally {
-      startStopLock.readLock().unlock();
+    } catch (InterruptedException ex) {
+      throw new ISE(ex, "failed to add [%s] lookup drop notice", lookupName);
     }
   }
 
-  /**
-   * Add or update a lookup factory
-   *
-   * @param lookupName             The name of the lookup
-   * @param lookupExtractorFactory The factory of the lookup
-   *
-   * @return True if the lookup was updated, false otherwise
-   *
-   * @throws IllegalStateException if start of the factory fails
-   */
-  public boolean updateIfNew(String lookupName, final LookupExtractorFactory lookupExtractorFactory)
+  @Nullable
+  public LookupExtractorFactoryContainer get(String lookupName)
   {
-    boolean update = false;
+    assertStarted();
+
+    startStopLock.readLock().lock();
     try {
-      startStopLock.readLock().lockInterruptibly();
-    }
-    catch (InterruptedException e) {
-      throw Throwables.propagate(e);
+      return lookupMap.get(lookupName);
+    } finally {
+      startStopLock.readLock().unlock();
     }
-    try {
-      assertStarted();
-      LookupExtractorFactory prior = lookupMap.get(lookupName);
-      update = lookupExtractorFactory.replaces(prior);
-      if (update) {
-        if (!lookupExtractorFactory.start()) {
-          throw new ISE("Could not start [%s]", lookupName);
-        }
-        boolean racy;
-        do {
-          if (prior == null) {
-            racy = null != lookupMap.putIfAbsent(lookupName, lookupExtractorFactory);
-          } else {
-            racy = !lookupMap.replace(lookupName, prior, lookupExtractorFactory);
-          }
+  }
 
-          if (racy) {
-            prior = lookupMap.get(lookupName);
-            update = lookupExtractorFactory.replaces(prior);
-          }
-        } while (racy && update);
+  public LookupsState getAllLookupsState()
```

Comment:
@leventov 
waiting on thread finish makes sense and will take care of the race condition. so i changed things to use LifecycleLock (as a result now this PR depends on LifecycleLock PR) and removed explicit locking.
however, using BlockingQueue still looks good to me so that part is not changed.